src/world/Units/Creatures/AIInterface.cpp

```diff
@@ -2477,3 +3091,546 @@ uint32 AIInterface::fixupCorridor(dtPolyRef* path, const uint32 npath, const uin
+
+uint32_t CreatureAISpells::getAttackStopTimer()
+{
+    return mAttackStopTimer;
+}
+
+void CreatureAISpells::setAnnouncement(std::string announcement)
+{
+    mAnnouncement = announcement;
+}
+
+void CreatureAISpells::sendAnnouncement(Unit* pUnit)
+{
+    if (!mAnnouncement.empty() && pUnit != nullptr)
+    {
+        sLogger.debug("AISpellEmotes::sendAnnouncement() : called");
+
+        pUnit->sendChatMessage(CHAT_MSG_RAID_BOSS_EMOTE, LANG_UNIVERSAL, mAnnouncement.c_str());
+    }
+}
+
+void CreatureAISpells::setCustomTarget(Unit* targetCreature)
+{
+    mCustomTargetCreature = targetCreature;
+}
+
+Unit* CreatureAISpells::getCustomTarget()
+{
+    return mCustomTargetCreature;
+}
+
+CreatureAISpells* AIInterface::addAISpell(uint32_t spellId, float castChance, uint32_t targetType, uint32_t duration /*= 0*/, uint32_t cooldown /*= 0*/, bool forceRemove /*= false*/, bool isTriggered /*= false*/)
+{
+    const auto spellInfo = sSpellMgr.getSpellInfo(spellId);
+    if (spellInfo != nullptr)
+    {
+        uint32_t spellDuration = duration * 1000;
+        if (spellDuration == 0)
+            spellDuration = spellInfo->getSpellDefaultDuration(nullptr);
+
+        uint32_t spellCooldown = cooldown * 1000;
+        if (spellCooldown == 0)
+            spellCooldown = spellInfo->getSpellDefaultDuration(nullptr);
+
+        CreatureAISpells* newAISpell = new CreatureAISpells(spellInfo, castChance, targetType, spellDuration, spellCooldown, forceRemove, isTriggered);
+
+        mCreatureAISpells.push_back(newAISpell);
+
+        newAISpell->setdurationTimer(spellDuration);
+        newAISpell->setCooldownTimer(spellCooldown);
+
+        return newAISpell;
+    }
+
+    sLogger.failure("tried to add invalid spell with id %u", spellId);
+
+    // assert spellInfo can not be nullptr!
+    ARCEMU_ASSERT(spellInfo != nullptr);
+    return nullptr;
+}
+
+void AIInterface::UpdateAISpells()
+{
+    if (mLastCastedSpell)
+    {
+        if (!mSpellWaitTimer.isTimePassed())
+        {
+            // spell has a min/max range
+            if (!getUnit()->isCastingSpell() && (mLastCastedSpell->mMaxPositionRangeToCast > 0.0f || mLastCastedSpell->mMinPositionRangeToCast > 0.0f))
+            {
+                // if we have a current target and spell is not triggered
+                if (mCurrentSpellTarget != nullptr && !mLastCastedSpell->mIsTriggered)
+                {
+                    // interrupt spell if we are not in  required range
+                    const float targetDistance = getUnit()->GetPosition().Distance2DSq({ mCurrentSpellTarget->GetPositionX(), mCurrentSpellTarget->GetPositionY() });
+                    if (!mLastCastedSpell->isDistanceInRange(targetDistance))
+                    {
+                        sLogger.debug("Target outside of spell range (%u)! Min: %f Max: %f, distance to Target: %f", mLastCastedSpell->mSpellInfo->getId(), mLastCastedSpell->mMinPositionRangeToCast, mLastCastedSpell->mMaxPositionRangeToCast, targetDistance);
```

Comment:
is this really a general debug message or more likely some debugFlag Scripts?

Reply:
debug flag